service/matching/matching_engine_test.go

```diff
@@ -2747,6 +2766,376 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
+	s.NoError(err)
+	pgMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+
+	s.EqualValues(taskCount*numWorkers, s.taskManager.getTaskCount(ptq))
+
+	// Checking the maxReadLevel with the value of task stored in db
+	maxTaskId, _ := s.taskManager.maxTaskID(ptq)
+	s.EqualValues(maxTaskId, pgMgr.backlogMgr.db.maxReadLevel.Load())
+
+	// validating the approximateBacklogCounter
+	s.EqualValues(taskCount*numWorkers, pgMgr.backlogMgr.db.getApproximateBacklogCount())
+
+	// Unloading the PQM
+	s.matchingEngine.unloadTaskQueuePartition(partitionManager)
+
+	// Simulating a TTL'ed task in Cassandra by removing it from the DB
+	// Remove the task from testTaskManager but not from db/AckManager
+
+	// Stopping the backlogManager so that we TTL and the taskReader does not catch this
+	request := &persistence.CompleteTasksLessThanRequest{
+		NamespaceID:        namespaceId,
+		TaskQueueName:      taskQueue.Name,
+		TaskType:           enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+		ExclusiveMaxTaskID: minTaskID + 1,
+		Limit:              100,
+	}
+	_, err = s.taskManager.CompleteTasksLessThan(context.Background(), request)
+	s.NoError(err)
+	s.EqualValues((taskCount*numWorkers)-1, s.taskManager.getTaskCount(ptq))
+
+	// adding pollers which shall also load the fresher version of tqm
+	s.pollWorkflowTasks(false, workflowType, 1, (taskCount*numWorkers)-1, ptq, taskQueue, nil)
+
+	// Updating pgMgr to have the latest pgMgr
+	pgMgr = s.getPhysicalTaskQueueManagerImpl(ptq)
+
+	// overwriting the maxReadLevel since it could have increased if the previous taskWriter was stopped (which would not result in resetting);
+	// This should never be called and is only being done here for test purposes
+	pgMgr.backlogMgr.db.SetMaxReadLevel(maxTaskId)
+
+	s.EqualValues(0, s.taskManager.getTaskCount(ptq))
+	s.Eventually(func() bool {
+		return int64(0) == pgMgr.backlogMgr.db.getApproximateBacklogCount()
+	}, 3*time.Second, 10*time.Millisecond, "backlog counter should have been reset")
+
+	s.EqualValues(int64(0), pgMgr.backlogMgr.db.getApproximateBacklogCount())
+}
+
+// TestResettingBacklogCounter tests the scenario where approximateBacklogCounter over-counts and resets it accordingly
+func (s *matchingEngineSuite) TestResetBacklogCounterNoDBErrors() {
+	s.resetBacklogCounter(2, 2, 2)
+}
+
+func (s *matchingEngineSuite) TestResetBacklogCounterDBErrors() {
+	s.taskManager.dbError = true
+	s.resetBacklogCounter(2, 2, 2)
+}
+
+func (s *matchingEngineSuite) TestMoreTasksResetBacklogCounterNoDBErrors() {
+	s.resetBacklogCounter(10, 20, 2)
+}
+
+func (s *matchingEngineSuite) TestMoreTasksResetBacklogCounterDBErrors() {
+	s.taskManager.dbError = true
+	s.resetBacklogCounter(10, 50, 5)
+}
+
+func (s *matchingEngineSuite) TestPersistAckLevelWithTaskCreationValidateBacklogCounter() {
+	// This test should fail since we are not persisting the ack level with every write.
+	s.T().Skip("Skipping this until AckLevel is persisted with task creation")
```

Comment:
This isn't ever going to happen though, right? Aren't we going in the other direction and want to stop writing any metadata on task creation, only verifying rangeid?

Reply:
yeah - i thought of still keeping this test in the event that we *may* come back to this but I can remove this for now.

Reply:
I keep a local directory of discarded code that I might want to come back to one day. Of course, no one else can find it there, probably including me in a week.

If you want to be sneaky, you can leave this in this PR and then remove it in the next PR, then it's in git history forever. Though probably no one will know to look for it.